docs/proposals/20210301-enhancement_of_YurtHub_caching_ability.md

```diff
@@ -0,0 +1,375 @@
+certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+coordinationv1 "k8s.io/api/coordination/v1"
+coordinationv1beta1 "k8s.io/api/coordination/v1beta1"
+corev1 "k8s.io/api/core/v1"
+discoveryv1alpha1 "k8s.io/api/discovery/v1alpha1"
+eventsv1beta1 "k8s.io/api/events/v1beta1"
+extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+networkingv1 "k8s.io/api/networking/v1"
+networkingv1beta1 "k8s.io/api/networking/v1beta1"
+nodev1alpha1 "k8s.io/api/node/v1alpha1"
+nodev1beta1 "k8s.io/api/node/v1beta1"
+policyv1beta1 "k8s.io/api/policy/v1beta1"
+rbacv1 "k8s.io/api/rbac/v1"
+rbacv1alpha1 "k8s.io/api/rbac/v1alpha1"
+rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+schedulingv1 "k8s.io/api/scheduling/v1"
+schedulingv1alpha1 "k8s.io/api/scheduling/v1alpha1"
+schedulingv1beta1 "k8s.io/api/scheduling/v1beta1"
+settingsv1alpha1 "k8s.io/api/settings/v1alpha1"
+storagev1 "k8s.io/api/storage/v1"
+storagev1alpha1 "k8s.io/api/storage/v1alpha1"
+storagev1beta1 "k8s.io/api/storage/v1beta1"
+v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+runtime "k8s.io/apimachinery/pkg/runtime"
+schema "k8s.io/apimachinery/pkg/runtime/schema"
+serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+```
+
+#### Problem One
+
+Because scheme currently only saves the above fixed resources' structures, when YurtHub receives Custom Resources returned from kube-apiserver, it cannot find the corresponding GO type structure to decode Custom Resources.
+
+### cacheManager
+
+CacheManager provides methods for managing resources cached on edge nodes, which are divided into three main functions:
+
+- Determining whether the response message needs to be cached
+- The resource byte stream from Kube-apiserver is decoded into `runtime.Object` and stored in memory. When the YurtHub operation is completed, the object is converted to a byte stream and stored on the hard disk
+
+- When cloud-edge communication is unhealthy, edge nodes use `QueryCache(...)` to find resources cached locally.
+
+```go
+// CacheManager is an adaptor to cache runtime object data into backend storage
+type CacheManager interface {
+	CacheResponse(ctx context.Context, prc io.ReadCloser, stopCh <-chan struct{}) error
+	QueryCache(req *http.Request) (runtime.Object, error)
+	UpdateCacheAgents(agents []string) error
+	ListCacheAgents() []string
+	CanCacheFor(req *http.Request) bool
+}
+```
+
+#### Problem Two
+
+Since the serializer currently used by YurtHub will lose GVK information when decoding resource objects. So we need to [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62) to obtain the corresponding kind information according to resource name. The two maps have the following three main functions:
```

Comment:
The two maps have the following three main functions --> These two maps have been used in the following three main functions.

Reply:
Proposal is already updated, thanks for review